Replace some hardcoded Aircraft assumptions with bool fields

reaperrr opened the pull request:
This lays some groundwork for easier activity merging (Fly & HeliFly etc.) and easier hybrid behavior implementation.

Additionally removed single-use `FlyCircleTimed` activity.

reaperrr committed:
Add TakeOffOnCreation and TakeOffOnResupply to Aircraft

Before this, it was impossible to replicate the behavior of the original games (staying on pad/airfield after reload) without hacking around in Mods.Common.
This allows modders to disable these without meddling with code.

reaperrr committed:
Remove FlyCircleTimed activity

A FlyCircle overload is sufficient.

reaperrr committed:
Upgrade rule for TakeOffOnResupply

reaperrr committed:
Upgrade rule for setting VTOL to true for CanHover actors

pchote commented:
I'd really prefer it if we can use multiple flags for each feature (`Hovers`, `VTOL`, etc) instead of having an enum with fixed characteristics.

reaperrr committed:
Introduce Aircraft VTOL boolean

Rather than hard-linking vertical take-off/land to the CanHover = Helicopter assumption.

reaperrr commented:
Merged #13638 into this.

reaperrr commented:
I used VTOL since nearly everybody knows what it means, and for those few who don't we have the description. I know we generally avoid abbreviations if possible, but that would make it quite long and a pain to write in this case.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyCircle.cs:
```diff
@@ -18,6 +18,13 @@ public class FlyCircle : Activity
 		readonly WDist cruiseAltitude;
+		int remainingTicks = -1;
+
+		public FlyCircle(Actor self, int ticks)
```
Implement this as a default parameter on the normal ctor instead: `public FlyCircle(Actor self int ticks = -1)`

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -187,7 +196,7 @@ public Aircraft(ActorInitializer init, AircraftInfo info)
 			// TODO: HACK: This is a hack until we can properly distinguish between airplane and helicopter!
 			// Or until the activities get unified enough so that it doesn't matter.
-			IsPlane = !info.CanHover;
+			IsPlane = !info.VTOL;
```
Can we remove this now?

reaperrr replied:
If you prefer to get rid of it now, I could replace `IsPlane` with `!info.VTOL` everywhere, yes. 
It just felt like bit of a wasted effort, since that would mean changing a lot of lines that will be removed entirely once we merge `Fly` with `HeliFly` etc., because most of these look-ups will shift from `Aircraft` to the respective activities then.

pchote replied:
Ok, fair enough.  I was thinking more that most of our plane activities could depend on `CanHover` and the landing on `VTOL` possibly without any further changes to the activities.  I may be wrong, however.

reaperrr replied:
Well, I went ahead and removed `IsPlane` afterall, it's not that much more work and it won't be much more work to shift the `if`s from the IMove activities to the merged activities later.

reaperrr committed:
Remove IsPlane hack

The new VTOL boolean together with CanHover is enough to replace this.

reaperrr commented:
Updated.

In case you wonder about it, I removed the commit that set `CanHover: true` on TS Banshee & Orca Bomber, it was bogus (I falsely assumed `CanHover` was triggering the `Hovers` effect).

pchote commented:
You can still add `VTOL: true` to those to give them proper takeoff/landing.

pchote commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -272,6 +272,7 @@
 		AirborneCondition: airborne
 		CruisingCondition: cruising
 		CanHover: True
+		VTOL: true
```
These should have `TakeOffOnResupply` too to retain legacy behaviour.

pchote reviewed (commented):
LGTM otherwise :+1:

reaperrr commented:
> You can still add `VTOL: true` to those to give them proper takeoff/landing.

I could, but the combination of `VTOL: true` and `CanHover: false` doesn't work very well yet. On take-off, it will ignore `VTOL: true` and just do a plane-like take-off, and when ordered to land, it will ignore `CanHover: false` and do a helicopter-like "move straight towards target while visually turning". 
Reason is that `TakeOff` just queues `MoveTo` which doesn't know when it's a take-off, and the only ways to fix that without regressions are either a) some very hacky temporary hacks, which I'd have to write first, or b) fixing it properly by merging the Fly and Land activities, for which I have a 95% finished local branch.
So I'd rather leave VTOL disabled on them for now and enable it with the next PR where it can serve as test-case for the merged activities.

reaperrr commented:
Updated.

Mailaender reviewed (approved):
Code looks sane :+1: and I can't spot regressions in-game. :white_check_mark:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b2b90d35d88ffb4603a64c9d7df83fd7164f6533...3c1b6c014e15c46faa9444763d7ca5793f9f9fa3)
